File: 1. myproject/6. eco-genaie/services/mcp-client-sdk/python/mcp_client.py
import grpc
import json
from typing import Dict, List, Optional

class MCPClient:
    def __init__(self, host: str = 'localhost', port: int = 9090):
        self.channel = grpc.insecure_channel(f'{host}:{port}')
        self.stub = None  # Will be initialized with generated gRPC stub

    def execute_tool(self, tool_name: str, parameters: Dict) -> Dict:
        """Execute a tool with given parameters."""
        request = {
            'tool_name': tool_name,
            'parameters': json.dumps(parameters)
        }
        response = self.stub.ExecuteTool(request)
        return json.loads(response.result) if response.success else None

    def get_tools(self) -> List[Dict]:
        """Get list of available tools."""
        request = {}
        response = self.stub.GetTools(request)
        return [self._convert_tool(tool) for tool in response.tools]

    def add_tool(self, tool: Dict) -> bool:
        """Add a new tool."""
        request = {'tool': tool}
        response = self.stub.AddTool(request)
        return response.success

    def _convert_tool(self, tool) -> Dict:
        """Convert gRPC tool message to dictionary."""
        return {
            'name': tool.name,
            'description': tool.description,
            'endpoint': tool.endpoint,
            'method': tool.method,
            'parameters': json.loads(tool.parameters)
        }

    def close(self):
        """Close the gRPC channel."""
        self.channel.close() 

File: 1. myproject/6. eco-genaie/services/mcp-client-sdk/java/src/main/java/com/yourorg/mcp/client/MCPClient.java
package com.yourorg.mcp.client;

import com.google.gson.Gson;
import com.yourorg.mcp.proto.*;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class MCPClient {
    private final ManagedChannel channel;
    private final MCPServiceGrpc.MCPServiceBlockingStub blockingStub;
    private final Gson gson = new Gson();

    public MCPClient(String host, int port) {
        this.channel = ManagedChannelBuilder.forAddress(host, port)
                .usePlaintext()
                .build();
        this.blockingStub = MCPServiceGrpc.newBlockingStub(channel);
    }

    public Map<String, Object> executeTool(String toolName, Map<String, Object> parameters) {
        ToolRequest request = ToolRequest.newBuilder()
                .setToolName(toolName)
                .setParameters(gson.toJson(parameters))
                .build();

        ToolResponse response = blockingStub.executeTool(request);
        if (!response.getSuccess()) {
            throw new RuntimeException(response.getError());
        }

        return gson.fromJson(response.getResult(), Map.class);
    }

    public List<Map<String, Object>> getTools() {
        GetToolsRequest request = GetToolsRequest.getDefaultInstance();
        GetToolsResponse response = blockingStub.getTools(request);

        return response.getToolsList().stream()
                .map(this::convertTool)
                .collect(Collectors.toList());
    }

    public boolean addTool(Map<String, Object> tool) {
        Tool toolProto = Tool.newBuilder()
                .setName((String) tool.get("name"))
                .setDescription((String) tool.get("description"))
                .setEndpoint((String) tool.get("endpoint"))
                .setMethod((String) tool.get("method"))
                .setParameters(gson.toJson(tool.get("parameters")))
                .build();

        AddToolRequest request = AddToolRequest.newBuilder()
                .setTool(toolProto)
                .build();

        AddToolResponse response = blockingStub.addTool(request);
        return response.getSuccess();
    }

    private Map<String, Object> convertTool(Tool tool) {
        return Map.of(
            "name", tool.getName(),
            "description", tool.getDescription(),
            "endpoint", tool.getEndpoint(),
            "method", tool.getMethod(),
            "parameters", gson.fromJson(tool.getParameters(), Map.class)
        );
    }

    public void shutdown() {
        channel.shutdown();
    }
} 